### PR TITLE
4.1

### PR DESCRIPTION
**Pull request by mrkno**
- sync middleware (with returns)
- run middleware
- match middleware
- common middleware
- jsdoc platform
- small bug fixes
- CI test fixes

### CODE REVIEW

File: Gruntfile.js [hunk 1]
```diff
@@ -52,9 +52,12 @@ module.exports = function (grunt) {
             concierge: {
                 cmd: 'node',
                 args: [
+                    '--use_strict',
+                    '--harmony',
                     'main.js',
+                    'grunt',
                     '--debug',
-                    'grunt'
+                    'silly'
```
Comment (fallaciousreasoning) [inline Gruntfile.js]: I mean......

File: core/common/arguments.js [hunk 2]
```diff
@@ -114,18 +114,18 @@ exports.parseArguments = (args, options, help = {enabled:false, string:null, col
             throw new Error('Invalid Arguments');
         }
 
-        const vals = [],
-            count = (pargs[0].expects || {}).length || 0;
+        const vals = [];
+        let count = (pargs[0].expects || {}).length || 0;
         for (let j = 1; j <= count; j++) {
```
Comment (fallaciousreasoning) [inline core/common/arguments.js]: Oh god, stuff is one indexed?

File: Gruntfile.js (hunk 1 above)
Comment (mrkno) [inline Gruntfile.js]: It is the lowest log level so I thought why not

File: core/common/arguments.js (hunk 2 above)
Comment (mrkno) [inline core/common/arguments.js]: Yup, current arg is at index 0, sub-arguments are 1+

File: core/platform.js [hunk 3]
```diff
@@ -7,14 +7,14 @@
 *
 * License:
 *        MIT License. All code unless otherwise specified is
-*        Copyright (c) Matthew Knox and Contributors 2015.
+*        Copyright (c) Matthew Knox and Contributors 2017.
```
Comment (fallaciousreasoning) [inline core/platform.js]: Was it really 2015 that we started? Wow.

Comment (mrkno) [inline core/platform.js]: 2014 actually, I think some of the (c) are still set to then too

File: core/platform.js [hunk 4]
```diff
@@ -7,14 +7,14 @@
...
 */
 
 const figlet = require('figlet'),
-    MiddlewareHandler = require(global.rootPathJoin('core/middleware.js'));
+    MiddlewareHandler = require('concierge/middleware');
 
 class Platform extends MiddlewareHandler {
-    constructor(bypassInit) {
+    constructor (bypassInit) {
```
Comment (fallaciousreasoning) [inline core/platform.js]: Is this ' ' convention?

File: core/platform.js [hunk 5]
```diff
@@ -73,26 +74,26 @@ class Platform extends MiddlewareHandler {
         return returnVal;
     }
 
-    _handleMessage(api, event) {
+    _handleMessage (api, event) {
```
Comment (fallaciousreasoning) [inline core/platform.js]: And here, I guess

File: core/platform.js (hunk 4 above)
Comment (mrkno) [inline core/platform.js]: There isn't a convention, I was just trying to make it more consistent

File: core/platform.js [hunk 6]
```diff
@@ -101,22 +102,44 @@ class Platform extends MiddlewareHandler {
         this.emitAsync('message', api, event);
     }
 
-    onMessage(api, event) {
+    /**
+     * Callback for integrations, for passing messages to integrations.
+     * @param {Object} api the api that raised the message.
+     * @param {Object} event the event that occured.
+     */
+    onMessage (api, event) {
```
Comment (fallaciousreasoning) [inline core/platform.js]: I'm just going to assume it is

File: core/platform.js [hunk 7]
```diff
@@ -101,22 +102,44 @@ class Platform extends MiddlewareHandler {
...
-    _loadSystemConfig() {
+    /**
+     * Gets a loaded module by name or filter function.
+     * @param {string|function()} arg either the name or a filter function to find the module.
+     * @return {Object} the module (if found), array-like otherwise.
+     * @emits Platform#started
+     */
+    getModule (arg) {
+        const modules = this.modulesLoader.getLoadedModules('module');
+        let func = arg;
```
Comment (fallaciousreasoning) [inline core/platform.js]: Ternary statement, maybe? Might be more readable your way though

File: core/platform.js [hunk 8]
```diff
@@ -175,7 +211,7 @@ class Platform extends MiddlewareHandler {
         // Unload user modules
         this.config.saveConfig();
         this.modulesLoader.unloadAllModules();
-        this.statusFlag = flag ? flag : global.StatusFlag.Shutdown;
+        this.statusFlag = flag || global.StatusFlag.Shutdown;
```
Comment (fallaciousreasoning) [inline core/platform.js]: This has got to be my favorite part of JS

File: test/acceptance/test_core_modules.js [hunk 9]
```diff
@@ -3,9 +3,10 @@ var assert = require('chai').assert,
     client = null;
 
 describe('Test core modules', function() {
-    this.timeout(5000);
+    this.timeout(10000);
```
Comment (fallaciousreasoning) [inline test/acceptance/test_core_modules.js]: Aaaah, the accepted way to make tests pass since ages ago :P

File: core/platform.js (hunk 8 above)
Comment (mrkno) [inline core/platform.js]: I agree, it is useful

File: test/acceptance/test_core_modules.js (hunk 9 above)
Comment (mrkno) [inline test/acceptance/test_core_modules.js]: In this case, it does make them pass, the timeout of 5s was too few

**Comment by fallaciousreasoning** [comment]
LGTM

File: core/common/arguments.js (hunk 2 above)
Comment (fallaciousreasoning) [inline core/common/arguments.js]: Oh right, that actually makes sense

File: core/platform.js (hunk 4 above)
Comment (fallaciousreasoning) [inline core/platform.js]: Fair enough

**Pull request merged**